### PR TITLE
Enable script cache on levelbuilder

### PR DESCRIPTION
**Pull request by balderdash**
We have a few different caches in various places, the one I'm turning on here is the one saved in class variables on models/script.rb: `@@script_cache`, `@@script_level_cache`, and `@level_cache`.

Basically this just changes `Script.should_cache?` to return true in levelbuilder mode, and adds `after_save` hooks to the script and level models to update the caches when needed. I also added a config option in locals.yml to explicitly enable/disable the cache, and turned on class caching in levelbuilder mode.

For a page like http://localhost.studio.code.org:3000/s/allthethings, this brings serving time down from 4.5s to 1s on my machine. Saving a modified script slowed from about 2s to 2.5s, but the reduced time spent rendering the modified script more than makes up for that lost time.

### CODE REVIEW

File: dashboard/app/models/script.rb [hunk 1]
```diff
@@ -158,7 +159,7 @@ def starting_level
 
   # Caching is disabled when editing scripts and levels or running unit tests.
   def self.should_cache?
-    return false if Rails.application.config.levelbuilder_mode
+    return false unless Rails.application.config.enable_script_cache
```
Comment (Hamms) [inline dashboard/app/models/script.rb]: I'm not sure I love the idea of caching defaulting to false; I'd much rather have things default to how we want them in production.

File: dashboard/app/views/scripts/show.html.haml [hunk 2]
```diff
@@ -66,7 +66,7 @@
 - if current_user && current_user.permission?(UserPermission::LEVELBUILDER)
   -# Show all the levels, their names, and instructions in the extra links box.
   = render layout: 'shared/extra_links' do
-    - levels = Script.get_from_cache(@script.name).stages.map{ |stage| stage.script_levels.map{ |sl| sl.level }}.flatten
+    - levels = Script.get_from_cache(@script.name).script_levels.map{ |sl| sl.levels }.flatten
```
Comment (Hamms) [inline dashboard/app/views/scripts/show.html.haml]: Does this correctly preserve ordering?

File: dashboard/app/models/script.rb [hunk 3]
```diff
@@ -228,6 +249,17 @@ def self.level_cache
...
+  def self.update_level_in_cache(level)
+    return unless self.should_cache?
+    level_cache[level.id] = level if level_cache.key? level.id
+    level_cache[level.name] = level if level_cache.key? level.name
+  end
+
+  def self.delete_level_from_cache(level)
+    return unless self.should_cache?
+    @@level_cache.except!(level.id, level.name)
+  end
```
Comment (Hamms) [inline dashboard/app/models/script.rb]: Why are these handled here and not on the level model?

File: dashboard/app/controllers/levels_controller.rb [hunk 4]
```diff
@@ -153,6 +153,7 @@ def create
   # DELETE /levels/1
   # DELETE /levels/1.json
   def destroy
+    Script.delete_level_from_cache(@level)
```
Comment (Hamms) [inline dashboard/app/controllers/levels_controller.rb]: Shouldn't this go in an `after_destroy` hook on the level model?

File: dashboard/app/models/script.rb (hunk 3 above)
Comment (Hamms) [reply dashboard/app/models/script.rb]: Oh, and do we need to remove the script from the cache when it is deleted, too?

File: dashboard/app/views/scripts/show.html.haml (hunk 2 above)
Comment (balderdash) [reply dashboard/app/views/scripts/show.html.haml]: Not sure, but https://github.com/code-dot-org/code-dot-org/pull/10928 made this change unnecessary anyway. Reverted except for changing sl.level to sl.levels to include any level variants.

File: dashboard/app/controllers/levels_controller.rb (hunk 4 above)
Comment (balderdash) [reply dashboard/app/controllers/levels_controller.rb]: Yup, done.

File: dashboard/app/models/script.rb (hunk 1 above)
Comment (balderdash) [reply dashboard/app/models/script.rb]: Changed the option to `disable_script_cache`.

File: dashboard/app/models/script.rb (hunk 3 above)
Comment (balderdash) [reply dashboard/app/models/script.rb]: Moved these methods, along with the other level cache and script level cache methods into their respective models. Also added an after_destroy handler for scripts.

**Comment by balderdash** [comment]
@Hamms PTAL

File: dashboard/app/models/level.rb [hunk 5]
```diff
@@ -162,6 +164,55 @@ def available_callouts(script_level)
...
+    return level unless level.nil?
+
+    # If the cache missed or we're in levelbuilder mode, fetch the level from
+    # the db. Note the field trickery is to allow passing an ID as a string,
+    # which some tests rely on (unsure about non-tests).
+    field = level_identifier.to_i.to_s == level_identifier.to_s ? :id : :name
+    level = Level.find_by!(field => level_identifier)
+    # Cache the level by ID and by name, unless it wasn't found.
+    @@level_cache[level.id] = level if level && Script.should_cache?
+    @@level_cache[level.name] = level if level && Script.should_cache?
```
Comment (Hamms) [inline dashboard/app/models/level.rb]: nit: could also just do `level.update_level_in_cache` here

**Comment by balderdash** [comment]
UI test failures are unrelated, merging.

**Pull request merged**